### PR TITLE
Ss 13 implementation

### PR DESCRIPTION
**Pull request by VolodymyrBilenko**

### CODE REVIEW

**Review by DanyloPlotskykh — APPROVED** [review]

File: _client/src/main.cpp [hunk 1]
```diff
@@ -10,7 +10,9 @@
 #include "Ui/Views/BaseView.hpp"
 #include "SpdlogConfig.hpp"
 
-const QString SERVER_API_URL{"http://127.0.0.1:8001/api"};
+#define SERVER_SCHEME   "http"
```
Comment (Sage-Cat) [inline _client/src/main.cpp]: we use `constexpr char[]` for this (you can check for reference `common/Endpoints.hpp`)
please prefer to use modern C++20 features, not defines

File: _client/src/main.cpp [hunk 2]
```diff
@@ -19,11 +21,13 @@ int main(int argc, char *argv[])
     SpdlogConfig::init<SpdlogConfig::LogLevel::Trace>();
     SPDLOG_INFO("SageStoreClient starting");
 
-    // Networking
-    NetworkService networkService;
-    networkService.setApiUrl(SERVER_API_URL);
-    networkService.setSerializer(std::make_unique<JsonSerializer>());
+    const NetworkService::ServerUrl serverUrl = { SERVER_SCHEME,
```
Comment (Sage-Cat) [inline _client/src/main.cpp]: ServerUrl serverUrl{.scheme = SERVER_PROTOCOL, .address =..., .port = ...};

File: _client/src/Network/NetworkService.hpp [hunk 3]
```diff
@@ -28,14 +28,20 @@ class NetworkService : public QObject
...
-    virtual void sendRequest(
-        QString endpoint,
-        Method method = Method::GET,
-        const Dataset &dataset = Dataset(),
-        const QString &resource_id = "");
-    void setApiUrl(const QString &apiUrl);
-    void setSerializer(std::unique_ptr<IDataSerializer> serializer);
+    typedef struct
+    {
+        QString scheme;
```
Comment (Sage-Cat) [inline _client/src/Network/NetworkService.hpp:33]: `protocol` is more appropriate name for our case

File: _client/src/Network/NetworkService.hpp [hunk 4]
```diff
@@ -28,14 +28,20 @@ class NetworkService : public QObject
...
+        QString address;
+        int     port;
+    } ServerUrl;
```
Comment (Sage-Cat) [inline _client/src/Network/NetworkService.hpp]: It's not actually a `ServerUrl` (cos when we say Url we expect it to be already merged string to address web resource), but a `ServerEndpoint` or `ServerConfig`

File: _client/src/Network/NetworkService.hpp [hunk 5]
```diff
@@ -28,14 +28,20 @@ class NetworkService : public QObject
...
 public:
-    explicit NetworkService(QObject *parent = nullptr);
-    virtual void sendRequest(
-        QString endpoint,
-        Method method = Method::GET,
-        const Dataset &dataset = Dataset(),
-        const QString &resource_id = "");
-    void setApiUrl(const QString &apiUrl);
-    void setSerializer(std::unique_ptr<IDataSerializer> serializer);
+    typedef struct
```
Comment (Sage-Cat) [inline _client/src/Network/NetworkService.hpp:31]: Why typedef? 

Just struct. It will improve readability

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Network/NetworkService.hpp (hunk 3 above)
Comment (Sage-Cat) [reply _client/src/Network/NetworkService.hpp:33]: Ah. I see, qt use name `scheme`, so it's fine

File: _client/src/Network/NetworkService.cpp [hunk 6]
```diff
@@ -1,29 +1,48 @@
...
+    m_serverUrl.setScheme(serverUrl.scheme);
+    m_serverUrl.setHost(serverUrl.address);
+    m_serverUrl.setPort(serverUrl.port);
+
+    m_serializer = std::move(serializer);
 }
 
 void NetworkService::sendRequest(QString endpoint, Method method, const Dataset &dataset, const QString &resource_id)
 {
+    QString serverPath;
```
Comment (Sage-Cat) [inline _client/src/Network/NetworkService.cpp]: it's not like `serverPath` but just `path` or `apiPath` or `endpointSuffix`

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Network/NetworkService.cpp [hunk 7]
```diff
@@ -1,29 +1,48 @@
 #include "NetworkService.hpp"
 #include "SpdlogConfig.hpp"
 
-NetworkService::NetworkService(QObject *parent)
-    : QObject(parent), m_manager(new QNetworkAccessManager(this))
+NetworkService::NetworkService(const ServerConfig &serverConfig, std::unique_ptr<IDataSerializer> serializer)
+    : m_manager(new QNetworkAccessManager(this))
```
Comment (Sage-Cat) [inline _client/src/Network/NetworkService.cpp]: `, m_serializer(std::move(serializer))`

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

**Review by Sage-Cat — APPROVED** [review]

**Pull request merged**